plugins/syntax/number.ml

```diff
@@ -0,0 +1,479 @@
+  let l, t = aux env [] t in
+  List.map (fun (na, a) -> na, EConstr.Unsafe.to_constr a) l,
+  EConstr.Unsafe.to_constr t
+
+(* [elaborate_to_post_params env sigma ty_ind params] builds the
+   [to_post] translation (c.f., interp/notation.mli) for the numeral
+   notation to parse/print type [ty_ind]. This translation is the
+   identity ([ToPostCopy]) except that it checks ([ToPostCheck]) that
+   the parameters of the inductive type [ty_ind] match the ones given
+   in [params]. *)
+let elaborate_to_post_params env sigma ty_ind params =
+  let to_post_for_constructor indc =
+    let sigma, c = match indc with
+      | GlobRef.ConstructRef c ->
+         let sigma,c = Evd.fresh_constructor_instance env sigma c in
+         sigma, Constr.mkConstructU c
+      | _ -> assert false in  (* c.f. get_constructors *)
+    let args, t = get_type env sigma c in
+    let params_indc = match Constr.kind t with
+      | Constr.App (_, a) -> Array.to_list a | _ -> [] in
+    let sz = List.length args in
+    let a = Array.make sz ToPostCopy in
+    if List.length params <> List.length params_indc then error_params ty_ind;
+    List.iter2 (fun param param_indc ->
+        match param, Constr.kind param_indc with
+        | Some p, Constr.Rel i when i <= sz -> a.(sz - i) <- ToPostCheck p
+        | _ -> ())
+      params params_indc;
+    indc, indc, Array.to_list a in
+  let pt_refs = get_constructors ty_ind in
+  let to_post_0 = List.map to_post_for_constructor pt_refs in
+  let to_post =
+    let only_copy (_, _, args) = List.for_all ((=) ToPostCopy) args in
+    if (List.for_all only_copy to_post_0) then [||] else [|to_post_0|] in
+  to_post, pt_refs
+
+(* [elaborate_to_post env sigma ty_name ty_ind l] builds the [to_post]
+   translation (c.f., interp/notation.mli) for the number notation to
+   parse/print type [ty_name] through the inductive [ty_ind] according
+   to the pairs [constant, constructor] in the list [l]. *)
+let elaborate_to_post_via env sigma ty_name ty_ind l =
+  let sigma, ty_name =
+    locate_global_sort_inductive_or_constant sigma ty_name in
+  let ty_ind = Constr.mkInd ty_ind in
+  (* Retrieve constants and constructors mappings and their type.
+     For each constant [cnst] and inductive constructor [indc] in [l], retrieve:
+     * its location: [lcnst] and [lindc]
+     * its GlobRef: [cnst] and [indc]
+     * its type: [tcnst] and [tindc] (decomposed in product by [get_type] above)
+     * [impls] are the implicit arguments of [cnst] *)
+  let l =
+    let read (consider_implicits, cnst, indc) =
+      let lcnst, lindc = cnst.CAst.loc, indc.CAst.loc in
+      let cnst, ccnst = locate_global_constructor_inductive_or_constant cnst in
+      let indc, cindc =
+        let indc = Smartlocate.global_constructor_with_alias indc in
+        GlobRef.ConstructRef indc, Constr.mkConstruct indc in
+      let get_type c =
+        (* ignore parameters of inductive types *)
+        let rm_params c = match Constr.kind c with
```

Comment:
You might want to whd_all here for the domains of the constructor types.

Reply:
I'm a bit confused why this example works fine except for emitting a warning; does it mean there's some extra work going on here that's only needed for emitting the warning, or have I not crafted a good enough example?
```coq
Module Test28.
Module Fin.
  Inductive t : nat -> Set :=
  | F1 {n : (nat : Set)} : (t (S n) : Set)
  | FS {n : (nat : Set)} : (t n : Set) -> (t (S n) : Set).
End Fin.

Inductive I :=
| I1 : I
| IS : I -> I.

Definition of_uint (x : Number.uint) : I :=
  let fix f n :=
      match n with
      | O => I1
      | S n => IS (f n)
      end in
  f (Nat.of_num_uint x).

Definition to_uint (x : I) : Number.uint :=
  let fix f i :=
      match i with
      | I1 => O
      | IS n => S (f n)
      end in
  Nat.to_num_uint (f x).

Local Open Scope type_scope.

Number Notation Fin.t via I using ([Fin.F1] I1, [Fin.FS] IS)
       of_uint to_uint : type_scope.
(* Warning: Type of IS seems incompatible with the type of Fin.FS.
Expected type is: ((Fin.t _UNBOUND_REL_1 : Set) -> I) instead of
(I -> I). This might yield ill typed terms when using the notation.
[via-type-mismatch,numbers]
*)

Check Fin.F1.
Check Fin.FS Fin.F1.
Check Fin.FS (Fin.FS Fin.F1).
Check Fin.FS (Fin.FS (Fin.FS Fin.F1)).
Check Fin.F1 : Fin.t 3.
Check Fin.FS Fin.F1 : Fin.t 3.
Check Fin.FS (Fin.FS Fin.F1) : Fin.t 3.
Fail Check Fin.FS (Fin.FS (Fin.FS Fin.F1)) : Fin.t 3.
Set Printing All.
Check 0.
Check 1.
Check 2.
Check 3.
Check 0 : Fin.t 3.
Check 1 : Fin.t 3.
Check 2 : Fin.t 3.
Fail Check 3 : Fin.t 3.
Unset Printing All.

End Test28.
```

Reply:
> does it mean there's some extra work going on here that's only needed for emitting the warning

Indeed, the types of `Fin.F1` and `Fin.FS` are used only to emit warnings. The notation itself is only based on the types of `I1` and `IS`.

> You might want to whd_all here for the domains of the constructor types.

Indeed, that fixed the warning.